File: config-service/src/utils/Encrypt.ts
import * as crypto from 'crypto';
import { SECRET_KEY } from '../domain/common/consts/Setup';
import { SystemException } from '../domain/exceptions/core/SystemException';

/**
 * Clase para manejar operaciones de encriptación AES-256-GCM y ofuscación base64
 * Basada en la implementación de encript.js
 * 
 * Esta clase implementa:
 * - AES-256-GCM para encriptación segura
 * - PBKDF2 para derivación de claves
 * - Ofuscación base64 con rotación de caracteres
 * - Compatibilidad con Web Crypto API
 */
export class Encrypt {

  private readonly password: string;

  constructor(password: string = SECRET_KEY) {
    /**
     * Inicializa el cifrador AES con una contraseña.
     * 
     * @param {string} password - Contraseña para derivar la clave de cifrado
     */
    this.password = password;
  }

  /**
   * Deriva una clave de 256 bits usando PBKDF2.
   * 
   * @param {Uint8Array} salt - Salt para la derivación de clave
   * @returns {Promise<CryptoKey>} Clave derivada
   */
  async _deriveKey(salt: Uint8Array) {
    const keyMaterial = await crypto.subtle.importKey(
      'raw',
      new TextEncoder().encode(this.password),
      'PBKDF2',
      false,
      ['deriveKey']
    );

    return await crypto.subtle.deriveKey(
      {
        name: 'PBKDF2',
        salt: salt,
        iterations: 100000,
        hash: 'SHA-256'
      },
      keyMaterial,
      {
        name: 'AES-GCM',
        length: 256
      },
      false,
      ['encrypt', 'decrypt']
    );
  }

  /**
   * Encripta el texto usando AES-256-GCM.
   * 
   * @param {string} plaintext - Texto a encriptar
   * @returns {Promise<string>} Texto encriptado codificado en base64
   */
  async encrypt(plaintext: string) {
    // Generar salt e IV aleatorios
    const salt = crypto.getRandomValues(new Uint8Array(16)); // 128 bits
    const iv = crypto.getRandomValues(new Uint8Array(12));   // 96 bits para GCM

    // Derivar clave
    const key = await this._deriveKey(salt);

    // Convertir texto a bytes
    const plaintextBytes = new TextEncoder().encode(plaintext);

    // Encriptar
    const encryptedData = await crypto.subtle.encrypt(
      {
        name: 'AES-GCM',
        iv: iv
      },
      key,
      plaintextBytes
    );

    // El resultado de AES-GCM incluye el tag al final
    const ciphertext = new Uint8Array(encryptedData);

    // Extraer ciphertext y tag (los últimos 16 bytes son el tag)
    const actualCiphertext = ciphertext.slice(0, -16);
    const tag = ciphertext.slice(-16);

    // Combinar: salt (16) + iv (12) + tag (16) + ciphertext
    const combined = new Uint8Array(salt.length + iv.length + tag.length + actualCiphertext.length);
    combined.set(salt, 0);
    combined.set(iv, 16);
    combined.set(tag, 28);
    combined.set(actualCiphertext, 44);

    // Codificar en base64
    return this._arrayBufferToBase64(combined);
  }

  /**
   * Desencripta el texto usando AES-256-GCM.
   * 
   * @param {string} encryptedData - Texto encriptado codificado en base64
   * @returns {Promise<string>} Texto desencriptado
   * @throws {Error} Si la desencriptación falla
   */
  async decrypt(encryptedData: string) {
    try {
      // Decodificar base64
      const data = this._base64ToArrayBuffer(encryptedData);

      // Extraer componentes
      const salt = data.slice(0, 16);           // Primeros 16 bytes
      const iv = data.slice(16, 28);           // Siguientes 12 bytes
      const tag = data.slice(28, 44);          // Siguientes 16 bytes
      const ciphertext = data.slice(44);       // Resto

      // Derivar clave
      const key = await this._deriveKey(salt);

      // Combinar ciphertext y tag para el formato que espera Web Crypto API
      const ciphertextWithTag = new Uint8Array(ciphertext.length + tag.length);
      ciphertextWithTag.set(ciphertext);
      ciphertextWithTag.set(tag, ciphertext.length);

      // Desencriptar
      const decryptedData = await crypto.subtle.decrypt(
        {
          name: 'AES-GCM',
          iv: iv
        },
        key,
        ciphertextWithTag
      );

      // Convertir bytes a texto
      return new TextDecoder().decode(decryptedData);

    } catch (error: any) {
      return new SystemException(`Error al desencriptar: ${error.message}`);
    }
  }

  /**
   * Convierte ArrayBuffer a string base64.
   * 
   * @param {Uint8Array} buffer - Buffer a convertir
   * @returns {string} String en base64
   */
  private _arrayBufferToBase64(buffer: Uint8Array) {
    let binary = '';
    const bytes = new Uint8Array(buffer);
    const len = bytes.byteLength;
    for (let i = 0; i < len; i++) {
      binary += String.fromCharCode(bytes[i]);
    }
    return btoa(binary);
  }

  /**
   * Convierte string base64 a Uint8Array.
   * 
   * @param {string} base64 - String en base64
   * @returns {Uint8Array} Array de bytes
   */
  private _base64ToArrayBuffer(base64: string) {
    const binaryString = atob(base64);
    const len = binaryString.length;
    const bytes = new Uint8Array(len);
    for (let i = 0; i < len; i++) {
      bytes[i] = binaryString.charCodeAt(i);
    }
    return bytes;
  }

  /**
   * Ofusca un string convirtiéndolo a base64 y aplicando una transformación adicional.
   * 
   * @param {string} texto - Texto a ofuscar
   * @returns {string} Texto ofuscado
   */
  ofuscarBase64(texto: string) {
    try {
      // Convertir a base64
      const textoBytes = new TextEncoder().encode(texto);
      const base64Texto = this._arrayBufferToBase64(textoBytes);

      // Aplicar transformación adicional: rotar caracteres
      const caracteres = base64Texto.split('');
      for (let i = 0; i < caracteres.length; i++) {
        if (/[a-zA-Z]/.test(caracteres[i])) {
          // Rotar letras: A->B, B->C, ..., Z->A
          if (caracteres[i] === caracteres[i].toUpperCase()) {
            caracteres[i] = String.fromCharCode(
              ((caracteres[i].charCodeAt(0) - 'A'.charCodeAt(0) + 1) % 26) + 'A'.charCodeAt(0)
            );
          } else {
            caracteres[i] = String.fromCharCode(
              ((caracteres[i].charCodeAt(0) - 'a'.charCodeAt(0) + 1) % 26) + 'a'.charCodeAt(0)
            );
          }
        } else if (/\d/.test(caracteres[i])) {
          // Rotar números: 0->1, 1->2, ..., 9->0
          caracteres[i] = ((parseInt(caracteres[i]) + 1) % 10).toString();
        }
      }

      return caracteres.join('');

    } catch (error: any) {
      throw new Error(`Error al ofuscar: ${error.message}`);
    }
  }

  /**
   * Desofusca un string que fue ofuscado con ofuscarBase64().
   * 
   * @param {string} textoOfuscado - Texto ofuscado a desofuscar
   * @returns {string} Texto original desofuscado
   * @throws {Error} Si la desofuscación falla
   */
  desofuscarBase64(textoOfuscado: string) {
    try {
      // Revertir la transformación: rotar caracteres en dirección opuesta
      const caracteres = textoOfuscado.split('');
      for (let i = 0; i < caracteres.length; i++) {
        if (/[a-zA-Z]/.test(caracteres[i])) {
          // Rotar letras en dirección opuesta: B->A, C->B, ..., A->Z
          if (caracteres[i] === caracteres[i].toUpperCase()) {
            caracteres[i] = String.fromCharCode(
              ((caracteres[i].charCodeAt(0) - 'A'.charCodeAt(0) - 1 + 26) % 26) + 'A'.charCodeAt(0)
            );
          } else {
            caracteres[i] = String.fromCharCode(
              ((caracteres[i].charCodeAt(0) - 'a'.charCodeAt(0) - 1 + 26) % 26) + 'a'.charCodeAt(0)
            );
          }
        } else if (/\d/.test(caracteres[i])) {
          // Rotar números en dirección opuesta: 1->0, 2->1, ..., 0->9
          caracteres[i] = ((parseInt(caracteres[i]) - 1 + 10) % 10).toString();
        }
      }

      const textoBase64 = caracteres.join('');

      // Decodificar base64
      const textoBytes = this._base64ToArrayBuffer(textoBase64);
      return new TextDecoder().decode(textoBytes);

    } catch (error: any) {
      throw new Error(`Error al desofuscar: ${error.message}`);
    }
  }



}

File: analysis-service/services/encrypt.py
import base64
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
import os

class Encrypt:
    def __init__(self, password):
        """
        Inicializa el cifrador AES con una contraseña.
        
        Args:
            password (str): Contraseña para derivar la clave de cifrado
        """
        self.password = password.encode('utf-8')
    
    def _derive_key(self, salt):
        """
        Deriva una clave de 256 bits usando PBKDF2.
        
        Args:
            salt (bytes): Salt para la derivación de clave
            
        Returns:
            bytes: Clave derivada de 32 bytes
        """
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,  # 256 bits
            salt=salt,
            iterations=100000,
        )
        return kdf.derive(self.password)
    
    def encrypt(self, plaintext):
        """
        Encripta el texto usando AES-256-GCM.
        
        Args:
            plaintext (str): Texto a encriptar
            
        Returns:
            str: Texto encriptado codificado en base64
        """
        # Generar salt e IV aleatorios
        salt = os.urandom(16)  # 128 bits
        iv = os.urandom(12)    # 96 bits para GCM
        
        # Derivar clave
        key = self._derive_key(salt)
        
        # Configurar cifrador
        cipher = Cipher(algorithms.AES(key), modes.GCM(iv))
        encryptor = cipher.encryptor()
        
        # Encriptar
        plaintext_bytes = plaintext.encode('utf-8')
        ciphertext = encryptor.update(plaintext_bytes) + encryptor.finalize()
        
        # Combinar: salt (16) + iv (12) + tag (16) + ciphertext
        encrypted_data = salt + iv + encryptor.tag + ciphertext
        
        # Codificar en base64
        return base64.b64encode(encrypted_data).decode('utf-8')
    
    def decrypt(self, encrypted_data):
        """
        Desencripta el texto usando AES-256-GCM.
        
        Args:
            encrypted_data (str): Texto encriptado codificado en base64
            
        Returns:
            str: Texto desencriptado
            
        Raises:
            Exception: Si la desencriptación falla
        """
        try:
            # Decodificar base64
            data = base64.b64decode(encrypted_data.encode('utf-8'))
            
            # Extraer componentes
            salt = data[:16]           # Primeros 16 bytes
            iv = data[16:28]          # Siguientes 12 bytes
            tag = data[28:44]         # Siguientes 16 bytes
            ciphertext = data[44:]    # Resto
            
            # Derivar clave
            key = self._derive_key(salt)
            
            # Configurar descifrador
            cipher = Cipher(algorithms.AES(key), modes.GCM(iv, tag))
            decryptor = cipher.decryptor()
            
            # Desencriptar
            plaintext_bytes = decryptor.update(ciphertext) + decryptor.finalize()
            
            return plaintext_bytes.decode('utf-8')
            
        except Exception as e:
            raise Exception(f"Error al desencriptar: {str(e)}")

    def ofuscar_base64(self, texto):
        """
        Ofusca un string convirtiéndolo a base64 y aplicando una transformación adicional.
        
        Args:
            texto (str): Texto a ofuscar
            
        Returns:
            str: Texto ofuscado
        """
        try:
            # Convertir a base64
            texto_bytes = texto.encode('utf-8')
            base64_texto = base64.b64encode(texto_bytes).decode('utf-8')
            
            # Aplicar transformación adicional: rotar caracteres
            caracteres = list(base64_texto)
            for i in range(len(caracteres)):
                if caracteres[i].isalpha():
                    # Rotar letras: A->B, B->C, ..., Z->A
                    if caracteres[i].isupper():
                        caracteres[i] = chr((ord(caracteres[i]) - ord('A') + 1) % 26 + ord('A'))
                    else:
                        caracteres[i] = chr((ord(caracteres[i]) - ord('a') + 1) % 26 + ord('a'))
                elif caracteres[i].isdigit():
                    # Rotar números: 0->1, 1->2, ..., 9->0
                    caracteres[i] = str((int(caracteres[i]) + 1) % 10)
            
            return ''.join(caracteres)
            
        except Exception as e:
            raise Exception(f"Error al ofuscar: {str(e)}")
    
    def desofuscar_base64(self, texto_ofuscado):
        """
        Desofusca un string que fue ofuscado con ofuscar_base64().
        
        Args:
            texto_ofuscado (str): Texto ofuscado a desofuscar
            
        Returns:
            str: Texto original desofuscado
            
        Raises:
            Exception: Si la desofuscación falla
        """
        try:
            # Revertir la transformación: rotar caracteres en dirección opuesta
            caracteres = list(texto_ofuscado)
            for i in range(len(caracteres)):
                if caracteres[i].isalpha():
                    # Rotar letras en dirección opuesta: B->A, C->B, ..., A->Z
                    if caracteres[i].isupper():
                        caracteres[i] = chr((ord(caracteres[i]) - ord('A') - 1) % 26 + ord('A'))
                    else:
                        caracteres[i] = chr((ord(caracteres[i]) - ord('a') - 1) % 26 + ord('a'))
                elif caracteres[i].isdigit():
                    # Rotar números en dirección opuesta: 1->0, 2->1, ..., 0->9
                    caracteres[i] = str((int(caracteres[i]) - 1) % 10)
            
            texto_base64 = ''.join(caracteres)
            
            # Decodificar base64
            texto_bytes = base64.b64decode(texto_base64.encode('utf-8'))
            return texto_bytes.decode('utf-8')
            
        except Exception as e:
            raise Exception(f"Error al desofuscar: {str(e)}")
